src/odbc/src/statement.cpp

```diff
@@ -994,6 +1007,73 @@ SqlResult::Type Statement::InternalClose() {
   return result;
 }
 
+void Statement::ExtendedFetch(SQLUSMALLINT orientation, SQLLEN offset, SQLULEN* rowCount, SQLUSMALLINT* rowStatusArray) {
+    IGNITE_ODBC_API_CALL(InternalExtendedFetch(orientation, offset, rowCount, rowStatusArray));
+}
+
+SqlResult::Type Statement::InternalExtendedFetch(SQLUSMALLINT orientation, SQLLEN offset,
+    SQLULEN* rowCount, SQLUSMALLINT* rowStatusArray) {
+  LOG_DEBUG_MSG("InternalExtendedFetch called with orientation " << orientation);
+  UNREFERENCED_PARAMETER(offset);
+
+  if (orientation != SQL_FETCH_NEXT) {
+    AddStatusRecord(SqlState::SHYC00_OPTIONAL_FEATURE_NOT_IMPLEMENTED,
+      "Only SQL_FETCH_NEXT FetchOrientation type is supported");
+    return SqlResult::AI_ERROR;
+  }
+
+  if (rowCount)
+    *rowCount = 0;
+
+  if (!currentQuery.get()) {
+    std::string errMsg = "Cursor is not in the open state.";
+    AddStatusRecord(SqlState::S24000_INVALID_CURSOR_STATE, errMsg);
+    return SqlResult::AI_ERROR;
+  }
+
+  SqlUlen columnBindOffsetValue = columnBindOffset ? *columnBindOffset : 0;
+  for (app::ColumnBindingMap::iterator it = columnBindings.begin();
+    it != columnBindings.end(); ++it) {
+    it->second.SetByteOffset(columnBindOffsetValue);
+  }
+
+  SQLINTEGER fetched = 0;
+  SQLINTEGER errors = 0;
+
+  LOG_DEBUG_MSG("rowsetSize is " << rowsetSize);
+  for (SqlUlen i = 0; i < rowsetSize; ++i) {
+    for (app::ColumnBindingMap::iterator it = columnBindings.begin();
+      it != columnBindings.end(); ++it) {
+      it->second.SetElementOffset(i);
+    }
+
+    SqlResult::Type res = currentQuery->FetchNextRow(columnBindings);
+
+    if (res == SqlResult::AI_SUCCESS || res == SqlResult::AI_SUCCESS_WITH_INFO) {
+      ++fetched;
+    } else if (res != SqlResult::AI_NO_DATA) {
+      ++errors;
+    }
+
+    if (rowStatusArray) {
+      rowStatusArray[i] = SqlResultToRowResult(res);
+    }
+  }
+
+  if (rowCount) {
+    *rowCount = fetched < 0 ? static_cast<SQLULEN>(rowsetSize) : fetched;
```

Comment:
Why does this get set to rowsetSize if fetched < 0?

Reply:
This is copied from `InternalFetchRow`. It may be some kind of way to handle overflows, as `fetched` is a `SQLINTEGER` in `InternalFetchRow`.

I have changed this to simply set `*rowCount` to `fetched`.

Reply:
I would check over the logic in InternalFetchRow to make sure it makes sense.